lms/services/digest.py

```diff
@@ -1,11 +1,303 @@
+    def __init__(self, db, audience, annotations):
+        self._db = db
+        self._audience = audience
+        self._annotations = annotations
+        self._unified_users = None
+        self._unified_courses = None
+
+    def instructor_digest(self, h_userid):
+        """
+        Return a digest (dict of template variables) for the given instructor.
+
+        The digest will only include courses in which both:
+
+        1. The user is an instructor
+        2. There are annotations by learners
+        """
+        course_digests = []
+
+        # Remove duplicates from self.unified_courses.values().
+        unified_courses = []
+        for unified_course in self.unified_courses.values():
+            if unified_course not in unified_courses:
+                unified_courses.append(unified_course)
+
+        for unified_course in unified_courses:
+            num_annotations = len(unified_course.learner_annotations)
+
+            if not num_annotations:
+                # There was no activity in this course.
+                continue
+
+            if h_userid not in [user.h_userid for user in unified_course.instructors]:
+                # The user isn't an instructor in this course.
+                continue
+
+            course_digests.append(
+                {"title": unified_course.title, "num_annotations": num_annotations}
+            )
+
+        return {
+            "total_annotations": sum(
+                course_digest["num_annotations"] for course_digest in course_digests
+            ),
+            "courses": course_digests,
+        }
+
+    @property
+    def unified_users(self):
+        """Return a dict mapping h_userid's to UnifiedUser's."""
+        if self._unified_users is not None:
+            return self._unified_users
+
+        self._unified_users = {}
+
+        h_userids = set(
+            self._audience
+            + [annotation["author"]["userid"] for annotation in self._annotations]
+        )
+
+        for h_userid in h_userids:
+            users = self._db.scalars(
+                select(User).filter_by(h_userid=h_userid)
+                # Order most-recently-updated first so that the code below will
+                # use the most recent values for email, display name, etc.
+                .order_by(User.updated.desc())
+            ).all()
+
+            email = None
+            for user in users:
+                if user.email:
+                    email = user.email
+                    break
+
+            display_name = None
+            for user in users:
+                if user.display_name:
+                    display_name = user.display_name
+                    break
+
+            self._unified_users[h_userid] = UnifiedUser(
+                h_userid=h_userid,
+                users=tuple(users),
+                email=email,
+                display_name=display_name,
+            )
+
+        return self._unified_users
+
+    @property
+    def unified_courses(self):
```

Comment:
I'm sure most if not all of this method could be written enterly in SQL (SQLA) but I'm not that motivated to attempt it to leave it here as feedback.

This seems much more contained in the number of queries anyway so making it just 1 or 2 is not that important. It wouldn't be a very readable one any way.

Reply:
Yeah, exactly. And I'm hoping it can be written to do a fixed number of queries too. Planning that for follow-up PRs